eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala

```diff
@@ -652,28 +652,35 @@ object Helpers {
       }
 
       // those are the preimages to existing received htlcs
-      val preimages = commitments.localChanges.all.collect { case u: UpdateFulfillHtlc => u.paymentPreimage }.map(r => Crypto.sha256(r) -> r).toMap
+      val preimages: Map[ByteVector32, ByteVector32] = commitments.localChanges.all.collect { case u: UpdateFulfillHtlc => u.paymentPreimage }.map(r => Crypto.sha256(r) -> r).toMap
+      val failedIncomingHtlcs: Set[Long] = commitments.localChanges.all.collect {
+        case u: UpdateFailHtlc => u.id
+        case u: UpdateFailMalformedHtlc => u.id
+      }.toSet
 
       val htlcTxs: Map[OutPoint, Option[HtlcTx]] = localCommit.htlcTxsAndRemoteSigs.collect {
         case HtlcTxAndRemoteSig(txInfo@HtlcSuccessTx(_, _, paymentHash, _, _), remoteSig) =>
           if (preimages.contains(paymentHash)) {
             // incoming htlc for which we have the preimage: we can spend it immediately
-            txInfo.input.outPoint -> withTxGenerationLog("htlc-success") {
+            Some(txInfo.input.outPoint -> withTxGenerationLog("htlc-success") {
               val localSig = keyManager.sign(txInfo, keyManager.htlcPoint(channelKeyPath), localPerCommitmentPoint, TxOwner.Local, commitmentFormat)
               Right(Transactions.addSigs(txInfo, localSig, remoteSig, preimages(paymentHash), commitmentFormat))
-            }
+            })
+          } else if (failedIncomingHtlcs.contains(txInfo.htlcId)) {
+            // incoming htlc that we know for sure will never be fulfilled downstream: we can safely discard it
+            None
```

Comment:
Good idea, it's a simpler way of filtering out unspendable htlcs without changing anything else, it's a good first step.
I find the use of `Option` and `flatten` a bit hard to read, we could leverage the `collect` and pattern matching instead:

```scala
      val htlcTxs: Map[OutPoint, Option[HtlcTx]] = localCommit.htlcTxsAndRemoteSigs.collect {
        // incoming htlc for which we have the preimage: we can spend it immediately
        case HtlcTxAndRemoteSig(txInfo@HtlcSuccessTx(_, _, paymentHash, _, _), remoteSig) if preimages.contains(paymentHash) =>
          txInfo.input.outPoint -> withTxGenerationLog("htlc-success") {
            val localSig = keyManager.sign(txInfo, keyManager.htlcPoint(channelKeyPath), localPerCommitmentPoint, TxOwner.Local, commitmentFormat)
            Right(Transactions.addSigs(txInfo, localSig, remoteSig, preimages(paymentHash), commitmentFormat))
          }
        // incoming htlc for which we don't have the preimage: we can't spend it immediately, but we may learn the
        // preimage later, otherwise it will eventually timeout and they will get their funds back
        // we ignore htlcs that have been failed downstream: we will never learn their preimage and the funds we may
        // have forwarded have already been refunded 
        case HtlcTxAndRemoteSig(txInfo: HtlcSuccessTx, _) if !failedIncomingHtlcs.contains(txInfo.htlcId) => txInfo.input.outPoint -> None
        case HtlcTxAndRemoteSig(txInfo: HtlcTimeoutTx, remoteSig) =>
          // outgoing htlc: they may or may not have the preimage, the only thing to do is try to get back our funds after timeout
          txInfo.input.outPoint -> withTxGenerationLog("htlc-timeout") {
            val localSig = keyManager.sign(txInfo, keyManager.htlcPoint(channelKeyPath), localPerCommitmentPoint, TxOwner.Local, commitmentFormat)
            Right(Transactions.addSigs(txInfo, localSig, remoteSig, commitmentFormat))
          }
      }.toMap
```

Not entirely sure whether it's better, feel free to chose the version you prefer.